src/linker/Linker/CompilerGeneratedState.cs

```diff
@@ -64,27 +139,43 @@ void PopulateCacheForType (TypeDefinition type)
 			return attribute.ConstructorArguments[0].Value as TypeDefinition;
 		}
 
-		public MethodDefinition? GetUserDefinedMethodForCompilerGeneratedMember (IMemberDefinition sourceMember)
+		// For state machine types/members, maps back to the state machine method.
+		// For local functions and lambdas, maps back to the owning method in user code (not the declaring
+		// lambda or local function, because the IL doesn't contain enough information to figure this out).
+		public bool TryGetOwningMethodForCompilerGeneratedMember (IMemberDefinition sourceMember, [NotNullWhen (true)] out MethodDefinition? owningMethod)
 		{
+			owningMethod = null;
 			if (sourceMember == null)
-				return null;
+				return false;
 
-			TypeDefinition compilerGeneratedType = (sourceMember as TypeDefinition) ?? sourceMember.DeclaringType;
-			if (_compilerGeneratedTypeToUserCodeMethod.TryGetValue (compilerGeneratedType, out MethodDefinition? userDefinedMethod))
-				return userDefinedMethod;
+			MethodDefinition? compilerGeneratedMethod = sourceMember as MethodDefinition;
+			if (compilerGeneratedMethod != null) {
+				if (_compilerGeneratedMethodToUserCodeMethod.TryGetValue (compilerGeneratedMethod, out owningMethod))
+					return true;
+			}
 
-			// Only handle async or iterator state machine
-			// So go to the declaring type and check if it's compiler generated (as a perf optimization)
-			if (!HasRoslynCompilerGeneratedName (compilerGeneratedType) || compilerGeneratedType.DeclaringType == null)
-				return null;
+			TypeDefinition sourceType = (sourceMember as TypeDefinition) ?? sourceMember.DeclaringType;
+
+			if (_compilerGeneratedTypeToUserCodeMethod.TryGetValue (sourceType, out owningMethod))
+				return true;
+
+			if (sourceType.DeclaringType == null)
+				return false;
+
+			var typeToCache = HasRoslynCompilerGeneratedName (sourceType) ? sourceType.DeclaringType : sourceType;
```

Comment:
If we need to check two levels deep in `PopulateCacheForType.ProcessMethod`, should this check if `sourceType.DeclaringType` is a compiler generated name and go up one more level if so?

Reply:
I think you are right, thanks! I couldn't find a good way to write a test for this because all the cases I can think of happen to cache the correct type before we get here. So I adjusted the logic and added an assert, hopefully this is good enough.

This also caused me to realize I had removed the check that only built a cache for compiler-generated types (it was actually building a cache for all types... oops).